kimchi/src/folding/mod.rs

```diff
@@ -0,0 +1,319 @@
+use self::error::{compute_error, Side};
+use crate::circuits::{
+    expr::{CacheId, Expr, FeatureFlag},
+    gate::GateType,
+};
+use ark_ff::{Field, One, Zero};
+use ark_poly::{EvaluationDomain, Evaluations, Radix2EvaluationDomain};
+use expressions::{folding_expression, IntegratedFoldingExpr};
+pub use instance_witness::{InstanceTrait, RelaxedInstance, RelaxedWitness, WitnessTrait};
+use instance_witness::{RelaxableInstance, RelaxablePair};
+use poly_commitment::PolyComm;
+use poly_commitment::{commitment::CommitmentCurve, SRS};
+use std::{collections::HashMap, marker::PhantomData, sync::Arc};
+
+mod error;
+mod expressions;
+mod instance_witness;
+#[cfg(test)]
+mod test;
+
+pub(crate) trait FoldingEnv<F> {
+    fn witness(&self, i: usize, side: Side) -> &Vec<F>;
+    fn witness_ext(&self, i: usize, side: Side) -> &Vec<F>;
```

Comment:
maybe use verbose for this one so it is easier to parse/find, given that the enum uses the verbose version of this (meaning, WitnessExtended)